File: spec/components/layouts/index_layout_spec.rb
# frozen_string_literal: true

require "spec_helper"

RSpec.describe Layouts::IndexLayout, feature_category: :design_system do
  let(:heading) { 'Page heading' }
  let(:description) { 'Page description' }
  let(:alerts_content) { 'Alert message' }
  let(:content) { 'Default content' }

  describe 'PageHeading' do
    describe 'heading' do
      it 'renders when heading prop is provided' do
        render_inline described_class.new(heading: heading)

        expect(page).to have_css('[data-testid="page-heading"]', text: heading)
      end

      it 'renders when heading slot is provided' do
        render_inline described_class.new do |c|
          c.with_heading { 'Custom Heading' }
        end

        expect(page).to have_css('[data-testid="page-heading"]', text: 'Custom Heading')
      end
    end

    describe 'description' do
      it 'renders description when prop provided' do
        render_inline described_class.new(heading: heading, description: description)

        expect(page).to have_css('[data-testid="page-heading-description"]', text: description)
      end

      it 'renders description when slot provided' do
        render_inline described_class.new(heading: heading) do |c|
          c.with_description { 'Test description' }
        end

        expect(page).to have_css('[data-testid="page-heading-description"]', text: 'Test description')
      end

      it 'does not render when no description prop or slot is provided' do
        render_inline described_class.new(heading: heading)

        expect(page).not_to have_css('[data-testid="page-heading-description"]')
      end
    end

    describe 'actions' do
      it 'renders actions when slot provided' do
        render_inline described_class.new(heading: heading) do |c|
          c.with_actions { 'Test action' }
        end

        expect(page).to have_css('[data-testid="page-heading-actions"]', text: 'Test action')
      end

      it 'does not render when no actions slot is provided' do
        render_inline described_class.new(heading: heading)

        expect(page).not_to have_css('[data-testid="page-heading-actions"]')
      end
    end
  end

  describe 'page_heading_sr_only' do
    it 'does not apply gl-sr-only class by default' do
      render_inline described_class.new(heading: heading)
      expect(page).not_to have_css('.gl-sr-only')
    end

    it 'applies gl-sr-only class when page_heading_sr_only is true' do
      render_inline described_class.new(heading: heading, page_heading_sr_only: true)
      expect(page).to have_css('.gl-sr-only')
    end
  end

  describe 'loading' do
    it 'renders spinner component when loading is true' do
      render_inline described_class.new(heading: heading, loading: true) do
        content
      end

      expect(page).to have_css('[data-testid="index-layout-loading-icon"]')
    end

    it 'does not render spinner component when loading is false' do
      render_inline described_class.new(heading: heading, loading: false) do
        content
      end

      expect(page).not_to have_css('[data-testid="index-layout-loading-icon"]')
    end

    it 'does not render content slot when loading' do
      render_inline described_class.new(heading: heading, loading: true) do
        content
      end

      expect(page).not_to have_css('[data-testid="index-layout-content"]', text: content)
    end

    it 'renders content slot when not loading' do
      render_inline described_class.new(heading: heading, loading: false) do
        content
      end

      expect(page).to have_css('[data-testid="index-layout-content"]', text: content)
    end
  end

  describe 'slots' do
    describe 'alerts' do
      it 'renders alert content when slot is provided' do
        render_inline described_class.new(heading: heading) do |c|
          c.with_alerts { alerts_content }
        end
        expect(page).to have_css('[data-testid="index-layout-alerts"]', text: alerts_content)
      end

      it 'does not render when no alerts slot is provided' do
        render_inline described_class.new(heading: heading)
        expect(page).not_to have_css('[data-testid="index-layout-alerts"]')
      end
    end

    describe 'default' do
      it 'renders body when default slot is provided' do
        render_inline described_class.new(heading: heading) do
          content
        end

        expect(page).to have_css('[data-testid="index-layout-content"]', text: content)
      end
    end
  end
end

File: spec/frontend/vue_shared/components/index_layout_spec.js
import { mount } from '@vue/test-utils';
import { GlLoadingIcon } from '@gitlab/ui';
import IndexLayout from '~/vue_shared/components/index_layout.vue';
import PageHeading from '~/vue_shared/components/page_heading.vue';

describe('IndexLayout', () => {
  let wrapper;

  const createComponent = (props = {}, slots = {}) => {
    wrapper = mount(IndexLayout, {
      propsData: props,
      slots,
    });
  };

  const findLoadingIcon = () => wrapper.findComponent(GlLoadingIcon);
  const findPageHeading = () => wrapper.findComponent(PageHeading);
  const findHeading = () => wrapper.find('[data-testid="page-heading"]');
  const findDescription = () => wrapper.find('[data-testid="page-heading-description"]');
  const findActions = () => wrapper.find('[data-testid="page-heading-actions"]');
  const findAlerts = () => wrapper.find('[data-testid="index-layout-alerts"]');
  const findContent = () => wrapper.find('[data-testid="index-layout-content"]');

  describe('PageHeading', () => {
    describe('heading', () => {
      it('renders when heading prop is provided', () => {
        createComponent({ heading: 'Test Heading' });
        expect(findPageHeading().exists()).toBe(true);
        expect(findPageHeading().props('heading')).toBe('Test Heading');
      });

      it('renders when heading slot is provided', () => {
        createComponent({}, { heading: 'Custom Heading' });
        expect(findHeading().exists()).toBe(true);
      });
    });

    describe('description', () => {
      it('renders description when prop provided', () => {
        createComponent({ heading: 'Test Heading', description: 'Test description' });
        expect(findDescription().exists()).toBe(true);
      });

      it('renders description when slot provided', () => {
        createComponent({ heading: 'Test Heading' }, { description: 'Test description' });
        expect(findDescription().exists()).toBe(true);
      });

      it('does not render when no description prop or slot is provided', () => {
        createComponent({ heading: 'Test Heading' });
        expect(findDescription().exists()).toBe(false);
      });
    });

    describe('actions', () => {
      it('renders actions when slot provided', () => {
        createComponent({ heading: 'Test Heading' }, { actions: 'Test action' });
        expect(findActions().exists()).toBe(true);
      });

      it('does not render when no actions slot is provided', () => {
        createComponent({ heading: 'Test Heading' });
        expect(findActions().exists()).toBe(false);
      });
    });
  });

  describe('headingTag', () => {
    it('defaults to null', () => {
      createComponent();
      expect(findPageHeading().props('headingTag')).toBeNull();
    });

    it('passes headingTag prop to PageHeading', () => {
      createComponent({ headingTag: 'h2' });
      expect(findPageHeading().props('headingTag')).toBe('h2');
    });
  });

  describe('pageHeadingSrOnly', () => {
    it('does not apply gl-sr-only class by default', () => {
      createComponent({ heading: 'Test Heading' });
      expect(findPageHeading().classes()).not.toContain('gl-sr-only');
    });

    it('applies gl-sr-only class when pageHeadingSrOnly is true', () => {
      createComponent({ heading: 'Test Heading', pageHeadingSrOnly: true });
      expect(findPageHeading().classes()).toContain('gl-sr-only');
    });
  });

  describe('loading', () => {
    it('does not render loading icon by default', () => {
      createComponent({ heading: 'Test Heading' });
      expect(findLoadingIcon().exists()).toBe(false);
    });

    it('renders loading icon when loading prop is true', () => {
      createComponent({ heading: 'Test Heading', loading: true });
      expect(findLoadingIcon().exists()).toBe(true);
    });

    it('does not render content slot when loading', () => {
      createComponent(
        { heading: 'Test Heading', loading: true },
        { default: '<div>Content</div>' },
      );
      expect(findLoadingIcon().exists()).toBe(true);
      expect(findContent().text()).not.toContain('Content');
    });

    it('renders content slot when not loading', () => {
      createComponent(
        { heading: 'Test Heading', loading: false },
        { default: '<div>Content</div>' },
      );
      expect(findLoadingIcon().exists()).toBe(false);
      expect(findContent().text()).toContain('Content');
    });
  });

  describe('slots', () => {
    describe('alerts', () => {
      it('renders alerts container when slot is provided', () => {
        createComponent({}, { alerts: '<div>Alerts slot content</div>' });
        expect(findAlerts().text()).toContain('Alerts slot content');
      });

      it('does not render when no alerts slot is provided', () => {
        createComponent({ heading: 'Test Heading' });
        expect(findAlerts().exists()).toBe(false);
      });
    });

    describe('default', () => {
      it('renders body when default slot is provided', () => {
        createComponent({}, { default: '<div>Content</div>' });
        expect(findContent().exists()).toBe(true);
      });
    });
  });
});
